fe/parameterize_batch_thermostat.py

```diff
@@ -0,0 +1,440 @@
+
+    host_pdb_file = args.complex_pdb
+    host_pdb = app.PDBFile(host_pdb_file)
+    host_conf = []
+    for x,y,z in host_pdb.positions:
+        host_conf.append([to_md_units(x),to_md_units(y),to_md_units(z)])
+    host_conf = np.array(host_conf)
+    host_name = "complex"
+
+    # set up the system
+    amber_ff = app.ForceField('amber99sbildn.xml', 'amber99_obc.xml')
+    host_system = amber_ff.createSystem(host_pdb.topology,
+        nonbondedMethod=app.NoCutoff,
+        constraints=None,
+        rigidWater=False)
+
+    cutoff = 1.25
+
+    host_system = openmm_converter.deserialize_system(host_system, cutoff=cutoff)
+    num_host_atoms = len(host_system.masses)
+
+    print("num_host_atoms", num_host_atoms)
+
+    for ligand_sdf_file in args.ligand_sdf:
+        for guest_mol in Chem.SDMolSupplier(ligand_sdf_file, removeHs=False):
+            break
+
+    open_ff = forcefield.Forcefield(args.forcefield)
+    nrg_fns = open_ff.parameterize(guest_mol, cutoff=cutoff)
+    guest_masses = get_masses(guest_mol)
+    guest_system = system.System(nrg_fns, open_ff.params, open_ff.param_groups, guest_masses)
+
+    combined_system = host_system.merge(guest_system)
+
+    temperature = 300
+    dt = 1.5e-3
+    friction = 91
+
+    ca, cbs, ccs = langevin_coefficients(
+        temperature,
+        dt,
+        friction,
+        np.array(combined_system.masses)
+    )
+
+    cbs *= -1
+
+    print("Integrator coefficients:")
+    print("ca", ca)
+    print("cbs", cbs)
+    print("ccs", ccs)
+
+    # insertion only
+    Ts = [
+        1000, # fast insertion
+        5000, # slow insertion
+    ]
+
+    complete_T = np.sum(Ts)
+
+    complete_lambda = np.concatenate([
+        np.linspace(cutoff, 0.3, Ts[0]),
+        np.linspace(0.3,   0.0,  Ts[1]),
+    ])
+
+    complete_cas = np.concatenate([
+        np.linspace(0.00, ca, Ts[0]),
+        np.linspace(ca, ca, Ts[1]),
+    ])
+
+    complete_dts = np.concatenate([
+        np.linspace(dt, dt, Ts[0]),
+        np.linspace(dt, dt, Ts[1]),
+    ])
+
+    lambda_idxs = np.zeros(len(combined_system.masses), dtype=np.int32)
+    lambda_idxs[num_host_atoms:] = 1
+
+    sim = simulation.Simulation(
+        combined_system,
+        complete_dts,
+        complete_cas,
+        cbs,
+        ccs,
+        complete_lambda,
+        lambda_idxs,
+        precision
+    )
+
+    initial_params = sim.system.params
+
+    lr = 1e-4
+    opt_init, opt_update, get_params = optimizers.adam(lr)
+
+    opt_state = opt_init(initial_params)
+
+    num_epochs = 90
+
+    # torsion, charge, gb radii, gb scaling factor, torsion
+    vary_allowed_groups = [(14, 1e-2)]*30 + [(12, 1e-2)]*30 + [(13, 1e-2)]*30
+
+    assert len(vary_allowed_groups) == num_epochs
+
+    ligand_list = []     
+
+    for ligand_sdf_file in args.ligand_sdf:
+        for rd_mol in Chem.SDMolSupplier(ligand_sdf_file, removeHs=False):
+            dG_value = rd_mol.GetProp("_dG")
+            mol_name = rd_mol.GetProp("s_user_Name")
```

Comment:
Something generally useful: if they don't encode the s_user_Name explicitly, you can use
```
rd_mol.GetProp('_Name')
```
To get the name (CHEMBL3402753_200) in the HEADER block off the sdf file:
```
CHEMBL3402753_200
                    3D
 Structure written by MMmdl.
 46 49  0  0  1  0            999 V2000
   24.6440   33.7566   50.5299 C   0  0  0  0  0  0
   24.1955   33.6079   51.9712 C   0  0  0  0  0  0
   22.9269   33.1279   52.2874 C   0  0  0  0  0  0
   22.5559   32.9819   53.6250 C   0  0  0  0  0  0
   23.4682   33.3170   54.6402 C   0  0  0  0  0  0
   24.7033   33.8055   54.3290 N   0  0  0  0  0  0
   25.0501   33.9349   53.0237 C   0  0  0  0  0  0
   23.1953   33.0558   56.0977 C   0  0  0  0  0  0
   24.2278   33.1076   57.0532 C   0  0  0  0  0  0
   24.0151   32.7150   58.3740 C   0  0  0  0  0  0
   22.7600   32.2678   58.7755 C   0  0  0  0  0  0
```

Reply:
In the hif2a ligands.sdf file they don't include the 's_user_Name' and call it the ID instead:

```
20 22  2  0  0  0
 20 27  1  0  0  0
 27 36  1  0  0  0
 27 37  1  0  0  0
M  END
> <ID>
338
```

Reply:
Yeah, that's why I use the shared "true_dG" I included as metric before, but if "_Name" works for all cases, I don't need the "ID" or "s_user_Name" or else.

Reply:
_Name should be always defined for a properly written SDF. @bp-kelley ?